### PR TITLE
custom vgg16 implementation

### PR DESCRIPTION
**Pull request by tristaaan**
This might be more accurate to the paper, however we cannot carry over the vgg16 weights which we were bfore. Despite this, the model performs almost identically!?

### CODE REVIEW

**Comment by tristaaan** [comment]
I think it's really mysterious how we can't seem to get under 0.026 and it always early stops in under 20 epochs. I really think it's my training size now.

**Comment by tristaaan** [comment]
```
Total             12178 # train and validation are this size
Training images   380 # total / 32
Validation images 380 # total / 32
```

So where are the other 10k or so images?

**Comment by tristaaan** [comment]
in `data_preprocessor` is this the right split? 

```
data_size = len(img_paths)
train_size      = int(data_size * 0.4)
validation_size = int(data_size * 0.4)
test_size       = int(data_size * 0.2)
```

Why do we care about test set here?

**Comment by tristaaan** [comment]
( @diegocasmo @ankurshukla03 )

**Comment by ankurshukla03** [comment]
Shouldn't in data_preprocess we should just split the data in training and validation ?
80-20%?

**Comment by tristaaan** [comment]
We shouldn't even be doing like this unless we care about reproducing results. I'm going to be sampling from the file list from now on with 80-20 split

**Comment by tristaaan** [comment]
mmmmmm there it is 
```
Traning images:   49262
Validtn images:   12316
Training batch:   1539
Validation batch: 384
```
Epochs take 151 seconds now.

**Comment by ankurshukla03** [comment]
so you are training on whole dataset.
kind of 2 hours if training for 50 epochs

**Comment by ankurshukla03** [comment]
We still need to fill the weekly plan

**Comment by tristaaan** [comment]
> so you are training on whole dataset.
> kind of 2 hours if training for 50 epochs

Seems so! Sticking around for an early stop, I'm not trusting it until I see some loss under .026 or we're past 20 epochs

File: code/custom-vgg16-gpu.ipynb [hunk 1]
```diff
@@ -143,54 +151,54 @@
...
-      "dense_5 (Dense)              (None, 4096)              16781312  \n",
+      "dense_2 (Dense)              (None, 4096)              16781312  \n",
       "_________________________________________________________________\n",
-      "dense_6 (Dense)              (None, 37)                151589    \n",
+      "dense_3 (Dense)              (None, 37)                151589    \n",
       "=================================================================\n",
-      "Total params: 50,526,053\n",
-      "Trainable params: 50,526,053\n",
+      "Total params: 39,194,405\n",
+      "Trainable params: 39,194,405\n",
```
Comment (tristaaan) [inline code/custom-vgg16-gpu.ipynb]: By reorganizing the input channels we require significantly fewer network parameters. This number here is down from 50.5 million where the input channel was organized 106,106,3. Here it's 3,106,106

**Comment by tristaaan** [comment]
The graph looks a bit better now, batches I believe are the right size and implemented properly. I’m really not sure why the batches are the size they are in the reference notebook (they’re 1?). In custom-vgg (no batch fix) I uploaded another solution csv last night and we’re performing a tad worse than the central pixel benchmark. So there is some progress I think it’s a good start.

**Comment by tristaaan** [comment]
It ran for the full 50 epochs, and we got to a loss of 0.1697. We can keep training this with the weights it generated if desired. It took approximately 12 hours to train which I think can be reduced a bit by fully preprocessing the images before hand so that the GPU is always working.

I ran the predictions through Kaggle and we now score **0.13070**. This would have put us at **153rd**. This is also better than the reference notebook we're basing this off of.

Repo-wise, I'm not quite sure how to reconcile the batch changes (move them to custom vgg and delete the batch-fix notebook?) I'm saving the weights interdependently so we can refer back to this model or train it more if desired.

**Comment by ankurshukla03** [comment]
Amazing, now we really got a good score.

> I ran the predictions through Kaggle and we now score 0.13070. This would have put us at 153rd. This is also better than the reference notebook we're basing this off of.

**Review by diegocasmo — COMMENTED** [review]
- As of now, we should only have 2 notebooks. The one that uses the pre-processed dataset to deal with less images (so that we can tell how parameters affect training without the need to use the whole dataset and submit to Kaggle), and then the notebook which uses the entire dataset and creates the submission ``.csv``
- The ``def preprocess_image(paths, i)`` should instead be ``def preprocess_image(img_path)``. No need for that method to be aware there's an image path array and find a particular index on it
- How does ``batch_size`` affect training? While ``32`` is suggested almost everywhere, what about if its ``16``, or ``64``? I expect the epoch time to be reduced/increased accordingly, but I'm most interested in checking whether it helps the weight updates or not (this might be easier to test in the notebook which does not use the entire dataset). Feel free to do this in another branch if desired.
  -  From the winner: ``I used a minibatch size of 16 examples.``
- This is the data pre-processing done by the winner of the competiton: ``[...] so I cropped all images to 207x207. I then downsampled them 3x to 69x69 [...]``. I wonder if we should change ours to match his?
- About training and validation set sizes (from the winner as well): ``For validation purposes, I split the training set in two parts. I used the first 90% for training, and the remainder for validation. ``
- Images are still not being rotated by a random multiple of  ``90`` (not sure how much this would help, but again, might be worth trying it in the notebook which uses the smaller dataset)
- Finally, for the sake of reproducibility, we should remove the randomization from the data processor notebook, that way we always train/validate/test with the same small dataset. This will allows us to better understand the hyper-parameters effect on learning.

**Comment by tristaaan** [comment]
> This is the data pre-processing done by the winner of the competition: [...] so I cropped all images to 207x207. I then downsampled them 3x to 69x69 [...]. I wonder if we should change ours to match his?

Wow, even more dimensionality reduction!

> Finally, for the sake of reproducibility, we should remove the randomization from the data processor notebook, that way we always train/validate/test with the same small dataset. This will allows us to better understand the hyper-parameters effect on learning.

Set a random seed before performing the split.

> From the winner: I used a minibatch size of 16 examples.

Let's do a run with batches of 16 then (#25 )

**Comment by tristaaan** [comment]
I'm thinking of altering this a bit: 

> As of now, we should only have 2 notebooks. [...]

to 

- [x] vgg script with a parameter to set the dataset size either as a fraction of the whole dataset or a fixed subset.
- [x] submission script

> The def preprocess_image(paths, i) should instead be def preprocess_image(img_path). No need for that method to be aware there's an image path array and find a particular index on it

This has been cleared up too. Amazing the code you can write when just fumbling around.

**Comment by diegocasmo** [comment]
@tristaaan 

Can you export your conda environment too? ``.yml`` file seems outdated

**Comment by tristaaan** [comment]
I exported what I'm using on my PC, it looks pretty different from the original one, but as long as conda can handle it. 

Caveat: I'm using `tensorflow-gpu` which is different from `tensorflow`

**Comment by tristaaan** [comment]
PTAL, some things are pretty reorganized.

**Review by diegocasmo — APPROVED** [review]
Have you trained with the newly preprocessed data? what result do we get (Kaggle)?

**Comment by tristaaan** [comment]
> Have you trained with the newly preprocessed data? what result do we get (Kaggle)?

Not yet, wanted to get this in order first. I'll continue this comment in #25
